src/traces/surface/defaults.js

```diff
@@ -29,12 +29,12 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
         return;
     }
 
-    traceOut._xlength = z[0].length;
-    traceOut._ylength = z.length;
-
-    coerce('x');
+    var x = coerce('x');
     coerce('y');
 
+    traceOut._xlength = (Array.isArray(x) && Array.isArray(x[0])) ? z.length : z[0].length;
+    traceOut._ylength = z.length;
```

Comment:
`_xlength` and `_ylength` are used in two places, to try and describe how much of each array to use - one in `surface/convert` (only for the `surface` type) and again in `gl3d/scene` (for all 3d types). The two do things in different ways, and especially since surface can use x/y arrays of either 1 or 2 dimensions, they weren't getting things right. I was trying to do better at handling mismatched arrays but this at least gets it right again for arrays of the correct length (and brings back the original `gl3d_ribbons` mock), and I've added an item for surface to #2353 to explore the mismatched cases more fully.